### PR TITLE
[Wasm GC] Add Monomorphize pass

### PR DESCRIPTION
**Pull request by kripken**
Monomorphization finds cases where we send more refined types to a function
than it declares. In such cases we can copy the function and refine the parameters:
```js
// B is a subtype of A

foo(new B());

function foo(x : A) { ..}

=>

foo_B(new B()); // call redirected to refined copy

function foo(x : A) { ..} // unchanged
function foo_B(x : B) { ..} // refined copy
```

This increases code size so it may not be worth it in all cases. This initial PR is
hopefully enough to start experimenting with this on performance, and so it does
not enable the pass by default.

Initial tests show this emits refined functions that look obviously better (fewer
casts), but I didn't see a benefit on the j2wasm benchmark I was running, which
might not happen to reach those paths. More work needed here.

This adds two variations of monomorphization, one that always does it, and the
default which is "careful": it sees whether monomorphizing lets the refined function
actually be better than the original (say, by removing a cast). If there is no
improvement then we do not make any changes. This saves a significant amount
of code size - on j2wasm the careful version increases by 13% instead of 20% -
but it does run more slowly obviously.

### CODE REVIEW

File: src/passes/Monomorphize.cpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+    // Note the list of all functions. We'll be adding more, and do not want to
+    // operate on those.
+    std::vector<Name> funcNames;
+    ModuleUtils::iterDefinedFunctions(
+      *module, [&](Function* func) { funcNames.push_back(func->name); });
+
+    // Find the calls in each function and optimize where we can, changing them
+    // to call more refined targets.
+    for (auto name : funcNames) {
```
Comment (tlively) [inline src/passes/Monomorphize.cpp:82]: We should probably go through and update the call sites (if not choose to generate further monomorphizations) in all of the new monomorphized functions as well.

**Review by tlively — COMMENTED** [review]
Very cool! It's nice to see how simple this is.

I wonder if there's a good way to detect that a function unconditionally downcasts its parameters, in which case we could replace the original function with a trampoline that downcasts and calls the specialized version of the function as a code size optimization.

**Comment by kripken** [comment]
> I wonder if there's a good way to detect that a function unconditionally downcasts its parameters, in which case we could replace the original function with a trampoline that downcasts and calls the specialized version of the function as a code size optimization.

Interesting, yeah, kind of a way to merge all the function aside from the first basic block.

File: src/passes/Monomorphize.cpp [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    // careful then we might change our mind.
+    auto chosenTarget = refinedTarget;
+    if (onlyWhenHelpful) {
+      // Optimize both functions using minimal opts, hopefully enough to see if
+      // there is a benefit to the refined types (such as the new types allowing
+      // a cast to be removed).
+      // TODO: Atm this can be done many times per function as it is once per
+      //       function and per set of types sent to it. Perhaps have some
+      //       total limit to avoid slow runtimes.
+      doMinimalOpts(func);
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp:158]: It might be worth considering. But the problem is that we don't actually have the property of "already fully optimized", not unless the user built with `--converge`, and not even then since that measures the entire module and not function by function (perhaps we should add `--converge-funcs`..?). In practice it is common that another `-O1` finds a few more bytes to save. And in those cases the refined function would win even if the refined types don't help at all. The only way really avoid that is to do an apples-to-apples comparison.

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/Monomorphize.cpp:158]: Hmm, in that case can we at least perform the minor optimizations (potentially to a fixed point) on the candidate functions all at once up front so they don't continue changing while we generate monomorphizations? But I still don't quite understand why it's not sufficient to run this optimization late in the user-configured pipeline so that things are "already fully optimized" according to whatever the user's definition of that is.

File: src/passes/Monomorphize.cpp [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    funcParamMap[{target, refinedParams}] = chosenTarget;
+
+    return chosenTarget;
+  }
+
+  // Run minimal function-level optimizations on a function.
+  void doMinimalOpts(Function* func) {
+    PassRunner runner(getPassRunner());
+    runner.options.optimizeLevel = 1;
+    runner.addDefaultFunctionOptimizationPasses();
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp:181]: Good idea, I'll add a TODO.

File: src/passes/Monomorphize.cpp [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+      if (costAfter >= costBefore) {
+        // We failed to improve. Remove the new function and return the old
+        // target.
+        module->removeFunction(refinedTarget);
+        chosenTarget = target;
+      }
+    }
+
+    // Mark the chosen target in the map, so we don't do this work again.
+    funcParamMap[{target, refinedParams}] = chosenTarget;
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp:172]: Yeah, I was wondering about this too... it seems like there are cases where each of those refinements is helpful and we want ones for B and C, but also cases where B and C help equally so we might as well use B for C, as you said. The vague idea I had was actually to do this in a later pass, like DuplicateFunctionElimination but for subtyping: when functions are identical aside from internal types, merge them. That's not simple though...

File: src/passes/Monomorphize.cpp (hunk 1 above)
Comment (tlively) [reply src/passes/Monomorphize.cpp:82]: Yeah, either in a subsequent loop over just the new functions or by continuously looping over all functions in the module until we stop making changes.

This is related to that TODO, but would also help less complicated cases where we just happen to optimize a caller before its callee so the monomorphized version of the caller doesn't call the optimized callee.

File: src/passes/Monomorphize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp:158]: > Hmm, in that case can we at least perform the minor optimizations (potentially to a fixed point) on the candidate functions all at once up front so they don't continue changing while we generate monomorphizations?

Maybe I'm missing your point here. I think you might be concerned about this situation: if `A` and `B` call `C`, then we would optimize `C` for purpose of `A->C` and also separately for `B->C`. So `C` gets optimized twice. That is necessary if the refined types differ, but if they match it's wasteful. However, we won't actually do it twice for matching types: once we see a pair of function and types we compute things once, store it in the cache `funcParamMap`, and reload from there. That includes the decision that it isn't worth monomorphizing - I'll clarify that in a comment.

Or are you worried about the main loop on functions getting to `A`, ending up optimizing `C`, then getting to `C`? That's fine, it will just look at more optimized code then. This is ordered and deterministic.

> But I still don't quite understand why it's not sufficient to run this optimization late in the user-configured pipeline so that things are "already fully optimized" according to whatever the user's definition of that is.

If a target function `T` has cost 100, and each subsequent run of `-O1` subtracts one, then it would take 100 full cycles to get to 0. `--converge` will do that (at least in a module with a single function), but generally we don't.

(And, since we don't, if we compare the cost of `T` to the cost of `T` with some refined types + `-O1` then the latter would look better just due to the opts even if refining doesn't help.)

File: src/passes/Monomorphize.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Monomorphize.cpp:85]: Makes sense. I'll add to the TODO.

**Review by aheejin — COMMENTED** [review]
Haven't read it all yet, but some questions?
- Do we remove the original functions in case they are not called by any functions? Or is there a separate pass that does it?
- In case there can be multiple versions of more refined params, for example, if A>B>C and the param is (A, A), there can be {(A, B), (B, C), (A, C), (C, A), ...}. Would it be more efficient to gather all values passed to the function, compute the LUB of them, and try to monomorphize?
- Can we possibly use the (existing) callgraph analysis to make the examination order more from the root towards leaves?

**Comment by kripken** [comment]
@aheejin 

> Do we remove the original functions in case they are not called by any functions? Or is there a separate pass that does it?

Yes, another pass does, `RemoveUnusedModuleElements` I think.

> In case there can be multiple versions of more refined params, for example, if A>B>C and the param is (A, A), there can be {(A, B), (B, C), (A, C), (C, A), ...}. Would it be more efficient to gather all values passed to the function, compute the LUB of them, and try to monomorphize?

Hmm, that might be worth trying if the pass is too slow, I'll add a TODO.

> Can we possibly use the (existing) callgraph analysis to make the examination order more from the root towards leaves?

Also a good idea - that does sound like it would be more effective in a single iteration. Also adding a TODO.

File: src/passes/Monomorphize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/Monomorphize.cpp:171]: I think @tlively is worried about us ending up optimizing `func` multiple times, namely per every possible `refinedFunc`.

[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp:171]: I see, thanks @aheejin , sorry for missing that.

I guess we could avoid that duplicated work if we saved the older versions and did a comparison using those, rather than the latest optimized code.

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/Monomorphize.cpp]: > Maybe I'm missing your point here. I think you might be concerned about this situation: if `A` and `B` call `C`, then we would optimize `C` for purpose of `A->C` and also separately for `B->C`. So `C` gets optimized twice.

Right, and the weird part here is that `C` has been optimized once when we decide whether to monomorphize the `A` callsite and twice when we decide whether to monomorphize the `B` callsite, so the two potential monomorphizations are not competing on equal footing. Our monomorphization decisions could change depending on the order in which we visit the `A` and `B` callsites. Yes, it's deterministic, but I would hope that our optimization decisions would be more path-independent. This will be more practically important if we ever try to parallelize some of this work.

Since each function has been optimized an arbitrary number of times and is arbitrarily far away from convergence at the point its cost is measured in the current code, it seems no less principled to take its initial state as the arbitrary point at which we measure its cost. This wouldn't be any more or less principled as a heuristic, but it could save us half the optimization work in this pass, and it also solves the path-dependence problem described above.

But I guess you're concerned about comparing the original function with N rounds of optimization to the monomorphized function with N+1 rounds of optimization? That makes sense, and I would solve it by comparing an optimized copy of the original function to a monomorphized + optimized copy of the function. To avoid holding multiple copies of every function at once, the top-level loop could be changed to loop first through callees (i.e. optimization candidates) rather than callers. It's a little more complicated, but I think this would be worth the effort. (It would be more cache-friendly, too.)

**Comment by tlively** [comment]
> > In case there can be multiple versions of more refined params, for example, if A>B>C and the param is (A, A), there can be {(A, B), (B, C), (A, C), (C, A), ...}. Would it be more efficient to gather all values passed to the function, compute the LUB of them, and try to monomorphize?
> 
> Hmm, that might be worth trying if the pass is too slow, I'll add a TODO.

A problem here is that the LUB may be the same as the original signature. For example if we have `B <: A` and `C <: A` and function `foo(A)` is called with `B` and `C` but not `A`, the LUB would be the same as the original signature. Maybe we could look at all the argument types and try to come up with some kind of minimal partitioning, though, and leave further partitioning to further runs of the pass. Another problem with that is that the minimal partitioning might not give enough benefits, though.

File: src/passes/Monomorphize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp]: > it seems no less principled to take its initial state as the arbitrary point at which we measure its cost. This wouldn't be any more or less principled as a heuristic

I think false positives are quite costly here, that is, cases where we monomorphize unnecessarily. That bloats the code for no good reason and could well end up making this pass appear unuseful. And that can easily happen each time there is even one tiny bit of optimization potential left in a function, if we wrongly assume it is already maximally optimized. So I think it is important to be 100% precise: only monomorphize when it helps.

> comparing an optimized copy of the original function to a monomorphized + optimized copy of the [original] function.

(`[original]` added by me, is that right?) IIUC it would always compare using the original function as a basis, and we'd immediately throw away any improvements we happen to find? I think that can work. However, throwing away even small improvements seems wasteful... Instead, I am leaning towards this recent TODO:

https://github.com/WebAssembly/binaryen/pull/5238/files#diff-e3faa07d688c017f0e69bcd2c24731585177e0c09d40a2034c1ca0111157f0b9R175-R178

Basically, if we optimize a function's cost and it's actually better, then we keep that improvement, and we'd optimize using that later. More work, but we've not thrown out good results. And, in the more common case, there will be no improvement to the cost, in which case we'd mark the function as "maximally optimized" and save that cost, then just use that saved cost later. Thoughts?

**Comment by aheejin** [comment]
> > > In case there can be multiple versions of more refined params, for example, if A>B>C and the param is (A, A), there can be {(A, B), (B, C), (A, C), (C, A), ...}. Would it be more efficient to gather all values passed to the function, compute the LUB of them, and try to monomorphize?
> > 
> > 
> > Hmm, that might be worth trying if the pass is too slow, I'll add a TODO.
> 
> A problem here is that the LUB may be the same as the original signature. For example if we have `B <: A` and `C <: A` and function `foo(A)` is called with `B` and `C` but not `A`, the LUB would be the same as the original signature. Maybe we could look at all the argument types and try to come up with some kind of minimal partitioning, though, and leave further partitioning to further runs of the pass. Another problem with that is that the minimal partitioning might not give enough benefits, though.

Hmm, also I guess the benefit this can achieve can be achieved anyway by `SignatureRefining` or `DeadArgumentElimination::refineArgumentTypes`..?

File: src/passes/Monomorphize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/Monomorphize.cpp]: > > it seems no less principled to take its initial state as the arbitrary point at which we measure its cost. This wouldn't be any more or less principled as a heuristic
> 
> I think false positives are quite costly here, that is, cases where we monomorphize unnecessarily... So I think it is important to be 100% precise: only monomorphize when it helps.

Yes, I agree. The solution I described would still only monomorphize when it helps. But it would do about half as much work to make that determination and the results wouldn't depend on the iteration order so it would be easier to parallelize.
 
> > comparing an optimized copy of the original function to a monomorphized + optimized copy of the [original] function.
> 
> (`[original]` added by me, is that right?) 

Yep 👍

> IIUC it would always compare using the original function as a basis, and we'd immediately throw away any improvements we happen to find? ...

No, we should replace the original function with the optimized copy once we no longer need the original version around to create new monomorphized copies of it.

> ... And, in the more common case, there will be no improvement to the cost, in which case we'd mark the function as "maximally optimized" and save that cost, then just use that saved cost later. Thoughts?

I think the "maximally optimized" idea is a red herring. The current code does not ensure or depend on anything being maximally optimized my suggested changes wouldn't change that. Detecting when a function is maximally optimized and caching its cost could save work, but not as much as my proposed changes would.

File: test/lit/passes/monomorphize.wast [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+    )
+  )
+)
+
+
+;; ALWAYS:      (func $refinable_0 (type $ref|$B|_=>_none) (param $ref (ref $B))
+;; ALWAYS-NEXT:  (drop
+;; ALWAYS-NEXT:   (local.get $ref)
+;; ALWAYS-NEXT:  )
+;; ALWAYS-NEXT: )
```
Comment (aheejin) [inline test/lit/passes/monomorphize.wast:122]: Hmm I think this should be within the module... Maybe a script bug.

Comment (kripken) [reply test/lit/passes/monomorphize.wast:122]: Yes, this looks a little odd, but new functions get up emitted after the module, for some reason, in the lit auto-update script.

Comment (tlively) [reply test/lit/passes/monomorphize.wast:122]: It's because the function's name doesn't match anything in the input module. so it gets lazily dumped before anything in the next module. In principle we could make this nicer by dumping it before the end of the module it comes from, but the script doesn't have detailed enough parse information to easily do that, IIRC.

File: src/passes/Monomorphize.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp]: > No, we should replace the original function with the optimized copy once we no longer need the original version around to create new monomorphized copies of it.

I see. That would only help once, though, since it's possible that we optimize `A` into `A'` which is better, and then `A'` can be further improved again to `A''`. That's probably rare though.

A larger concern with saving the original function is that it takes work, since we couldn't just optimize `A` in-place like we normally do, we'd need to first copy `A` and save that copy. In terms of memory usage, it would be better to avoid that, which the suggestion to note which functions are "maximally optimizable" would do (it just stores a bool+Index per function, not a copy of the body).

> I think the "maximally optimized" idea is a red herring. The current code does not ensure or depend on anything being maximally optimized my suggested changes wouldn't change that. Detecting when a function is maximally optimized and caching its cost could save work, but not as much as my proposed changes would.

Fair points, correct. Noting "maximally optimizable" functions would save some but not all the optimization work. It would save the copying and memory usage, though, and it would allow maximal improvement (`A''` from before).

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/Monomorphize.cpp]: > That would only help once, though, since it's possible that we optimize `A` into `A'` which is better, and then `A'` can be further improved again to `A''`. That's probably rare though.

And it shouldn't be the responsibility of this pass to do this extra optimization if it is possible. Since the current implementation happens to do repeated optimizations, it makes sense to persist those optimizations, but it would be better to prioritize the efficiency of the implementation and let the overall pass pipeline perform any additional optimizations.

> A larger concern with saving the original function is that it takes work, since we couldn't just optimize `A` in-place like we normally do, we'd need to first copy `A` and save that copy. In terms of memory usage...

Copying the function once seems like a trivial amount of work compared to running the full default function optimization pipeline on it. If we take care to only keep one of these function copies in memory at a time, the memory overhead also doesn't seem problematic. I guess with our arena allocation, the memory would never actually be freed, but that could be solved by allocating the copy in a temporary arena. Maybe all the candidate monomorphised functions should be held in temporary arenas until we decide to definitely include them, at which point they can be copied back into the main module?

[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/Monomorphize.cpp]: Copying is not much work, agreed. But the memory usage is more significant I think. We could explore ways to reduce the cost but the only reasonable idea I have is to make a new arena for all these functions, which would mean it lives til the end of the pass. That could be a significant amount of memory in some cases.

I agree it shouldn't be the responsibility of this pass to do incidental optimizations. But it is just simpler to do them and keep them.

I agree there is a downside to what I propose, in that it will do somewhat more optimization work than needed. But in the common case (the function is already fully optimized) it will do 0 extra work. And it keeps more incidental optimizations it happens to find.

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/Monomorphize.cpp]: Talked offline and I didn't realize how fast running the O1 optimization pipeline was. With a comment about the path-dependence property, this implementation LGTM for experimentation.

We also discussed some other potential future improvements such as monomorphizing for specific values in addition to refined types.

**Comment by kripken** [comment]
I added some comments now in the code after offline discussion. Overall, I think @tlively and I agree it seems best to leave optimizing how fast this pass runs for later, as atm it is fast enough for experimentation, and there are several possible paths to speed it up but none is obviously best.

Concretely, this takes a similar amount of time as e.g. `--gufa`, both take around 6 seconds on the Dart Barista benchmark for example. Good enough for now in both cases, I think, but eventually we'll want better.

**Review by aheejin — APPROVED** [review]

File: src/passes/Monomorphize.cpp (hunk 1 above)
Comment (tlively) [reply src/passes/Monomorphize.cpp:98]: Did we end up with a TODO for this? I don't see one now, but maybe I'm missing it.

**Review by tlively — APPROVED** [review]

File: src/passes/Monomorphize.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/Monomorphize.cpp:98]: I think the TODO on "multiple cycles" covers it, unless there is something I haven't understood from what you've said here?

https://github.com/WebAssembly/binaryen/pull/5238/files#diff-e3faa07d688c017f0e69bcd2c24731585177e0c09d40a2034c1ca0111157f0b9R37

Comment (kripken) [reply src/passes/Monomorphize.cpp:98]: Landing this, but let me know if there's a TODO I should add later.

Comment (tlively) [reply src/passes/Monomorphize.cpp:98]: Ah, solving the cycle problem would solve this problem, too, but solving this problem doesn't require solving the cycle problem. But I guess between the TODOs to solve the cycle problem and to visit functions in a more intelligent order, we probably don't need another separate TODO for this.
